packages/firestore/test/integration/api/array_transforms.test.ts

```diff
@@ -0,0 +1,253 @@
+import * as firestore from '@firebase/firestore-types';
+
+import firebase from '../util/firebase_export';
+import { apiDescribe, withTestDoc, withTestDb } from '../util/helpers';
+import { EventsAccumulator } from '../util/events_accumulator';
+
+// TODO(array-features): Remove "as any"
+// tslint:disable-next-line:no-any variable-name Type alias can be capitalized.
+const FieldValue = firebase.firestore.FieldValue as any;
+
+/**
+ * Note: Transforms are tested pretty thoroughly in server_timestamp.test.ts
+ * (via set, update, transactions, nested in documents, multiple transforms
+ * together, etc.) and so these tests mostly focus on the array transform
+ * semantics.
+ */
+// TODO(array-features): Enable these tests once the feature is available in
+// prod. For now you can run these tests using:
+// yarn test:browser --integration --local
+apiDescribe.skip('Array Transforms:', persistence => {
+  // A document reference to read and write to.
+  let docRef: firestore.DocumentReference;
+
+  // Accumulator used to capture events during the test.
+  let accumulator: EventsAccumulator<firestore.DocumentSnapshot>;
+
+  // Listener registration for a listener maintained during the course of the
+  // test.
+  let listenerRegistration: () => void;
+
+  /** Waits for a latency compensated local snapshot. */
+  async function waitForLocalEvent(): Promise<firestore.DocumentSnapshot> {
+    const snapshot = await accumulator.awaitEvent();
+    if (snapshot.metadata.hasPendingWrites) {
+      return snapshot;
+    } else {
+      return waitForLocalEvent();
+    }
+  }
+
+  /** Waits for a snapshot that has no pending writes */
+  async function waitForRemoteEvent(): Promise<firestore.DocumentSnapshot> {
+    const snapshot = await accumulator.awaitEvent();
+    if (!snapshot.metadata.hasPendingWrites) {
+      return snapshot;
+    } else {
+      return waitForRemoteEvent();
+    }
+  }
+
+  /** Writes some initialData and consumes the events generated. */
+  async function writeInitialData(
+    initialData: firestore.DocumentData
+  ): Promise<void> {
+    await docRef.set(initialData);
+    await waitForLocalEvent();
+    const snapshot = await waitForRemoteEvent();
+    expect(snapshot.data()).to.deep.equal(initialData);
+  }
+
+  async function expectLocalAndRemoteEvent(
+    expected: firestore.DocumentData
+  ): Promise<void> {
+    const localSnap = await waitForLocalEvent();
+    expect(localSnap.data()).to.deep.equal(expected);
+    const remoteSnap = await waitForRemoteEvent();
+    expect(remoteSnap.data()).to.deep.equal(expected);
+  }
+
+  /**
+   * Wraps a test, getting a docRef and event accumulator, and cleaning them
+   * up when done.
+   */
+  async function withTestSetup<T>(test: () => Promise<T>): Promise<void> {
+    await withTestDoc(persistence, async doc => {
+      docRef = doc;
+      accumulator = new EventsAccumulator<firestore.DocumentSnapshot>();
+      listenerRegistration = docRef.onSnapshot(
```

Comment:
Nit: While you are moving this, I feel that in JS "unsubscribe" is a much better name for this variable.